### PR TITLE
kernel: Make thread 'init_delay' k_timeout_t rather than int msecs

### PR DESCRIPTION
**Pull request by keith-packard**
Storing this value in milliseconds rather than using k_timeout_t requires the system to perform division at runtime to convert types. This pulls in the 64-bit soft division code on platforms without hardware for this.

Perform the conversion at build time instead by using the runtime time directly.

### CODE REVIEW

**Comment by keith-packard** [comment]
Note that to make this PR actually build, the computations in time_units.h probably need to be converted from inline functions to macros so that the constant values they generate can be used in static initializers for global data.

**Review by cfriedt — DISMISSED** [review]

**Comment by keith-packard** [comment]
ok, I've created a horror show of C macros which should make this all work. That's required because global variable initializers must be constants, and even though the static inline functions generated fixed values, they aren't "constant enough" for the compiler to allow them in those contexts. Welcome to the dark side of C macros though; I tried to make them as pretty as possible, but there's only so much that can be done ...

**Comment by keith-packard** [comment]
oh, and there's one really ugly kludge left -- i had to leave the size of the init_delay variable set to 4 bytes or none of the tests for the m5stack_core2 would link. Something something overlapping linker segments. Anyone understand that platform enough to help out?

**Comment by keith-packard** [comment]
The compliance checks are failing because of the macro formatting. I'd like to leave things as-is -- getting complex macros formatted so that people have a chance of understanding them seems more important than following the rules designed to make automatic formatting easier.

**Comment by keith-packard** [comment]
> oh, and there's one really ugly kludge left -- i had to leave the size of the init_delay variable set to 4 bytes or none of the tests for the m5stack_core2 would link. Something something overlapping linker segments. Anyone understand that platform enough to help out?

Turns out 'esptool' was ignoring alignment when merging ELF sections -- the 64-bit timer value caused the static thread init sections to require 8-byte instead of 4-byte alignment. PR pending at https://github.com/zephyrproject-rtos/hal_espressif/pull/243

**Comment by kartben** [comment]
Adding @fgrandel as IIRC he's been looking at similar build time conversions in #60400.

**Comment by keith-packard** [comment]
Here's some data motivating this series. Using picolibc 1.8.4 (which doesn't use 64-bit divmod). Building samples/hello_world for qemu_cortex_m3.
Before:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:        8946 B       256 KB      3.41%
             RAM:        3988 B        64 KB      6.09%
        IDT_LIST:          0 GB         2 KB      0.00%
```
After:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:        8198 B       256 KB      3.13%
             RAM:        3988 B        64 KB      6.09%
        IDT_LIST:          0 GB         2 KB      0.00%
```
A 748 byte savings in flash.

File: include/zephyr/sys/time_units.h [hunk 1]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
+/* Convert, generating a 32-bit result */
+#define z_tmcvt_32(__t, __from_hz, __to_hz, __const_hz, __round_up, __round_off) \
+	((__const_hz) ?							\
+	 (								\
+		 z_tmcvt_identity(__from_hz, __to_hz) ?			\
+		 (uint32_t) (__t)					\
+		 :							\
+		 z_tmcvt_div_ratio(__from_hz, __to_hz) ?		\
+		 (							\
+			 (int64_t) (__t) < (int64_t) BIT64(32) -	\
```
Comment (fgrandel) [inline include/zephyr/sys/time_units.h:189]: This is really hard to follow now. The original approach was much easier to read. Don't have an immediate idea how to fix it but looking again might be worth it?

I think the general approach of having two kinds of macros per conversion algorithm is the best way to go:
- one macro that checks applicability of a certain algorithm
- one macro that applies the algorithm

These macros can be named and placed in pairs such that their correspondence is obvious and easy to follow.

Then the basic functional approach to the conversion is `is_a_applicable(...) ? convert_by_a() : is_b_applicable(...) : convert_by_b() ... : convert_by_n()`.

This was the idea all the time anyway but now it can be made obvious and trivial rather than trying to be overly clever. Just my personal taste of course...

**Review by ghost — COMMENTED** [review]
Some nitpicks, some ideas, nothing blocking at all. Just notes I took trying to understand what's going on.

**Comment by fgrandel** [comment]
> oh, and there's one really ugly kludge left -- i had to leave the size of the init_delay variable set to 4 bytes or none of the tests for the m5stack_core2 would link. Something something overlapping linker segments. Anyone understand that platform enough to help out?

https://github.com/zephyrproject-rtos/zephyr/issues/58080 maybe?

**Comment by keith-packard** [comment]
> Some nitpicks, some ideas, nothing blocking at all. Just notes I took trying to understand what's going on.

wow, thanks for your detailed review. I'll try to answer the individual notes.

File: include/zephyr/sys/time_units.h [hunk 2]
```diff
@@ -259,11 +318,10 @@ static TIME_CONSTEXPR ALWAYS_INLINE uint64_t z_tmcvt(uint64_t t, uint32_t from_h
  *
  * @return The converted time value
  */
-static TIME_CONSTEXPR inline uint32_t k_ms_to_cyc_floor32(uint32_t t)
-{
-	/* Generated.  Do not edit.  See above. */
-	return z_tmcvt(t, Z_HZ_ms, Z_HZ_cyc, Z_CCYC, true, false, false);
-}
+/* Generated.  Do not edit.  See above. */
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:321]: Something like this?
```
/** @brief Convert ticks to hardware cycles. 64 bits. Rounds up.
 *
 * Converts time values in ticks to hardware cycles.
 * Computes result in 64 bit precision.
 * Rounds up to the next highest output unit.
 *
 * @param t Source time in ticks. uint64_t
 *
 * @return The converted time value in hardware cycles. uint64_t
 */
```

File: include/zephyr/sys/time_units.h [hunk 3]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
-			/* Slower algorithm but input is first divided before being multiplied
-			 * which prevents overflow of intermediate value.
-			 */
-			return (t / from_hz) * to_hz + ((t % from_hz) * to_hz + off) / from_hz;
-		}
-	}
-}
+
+/* true if the conversion is the identity */
+#define z_tmcvt_identity(__from_hz, __to_hz) \
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:120]: See above for some more verbose doxygen comments.

 As to types, the general approach I took was to always cast the incoming time value to uint64_t as that matches the original inline parameter type and should end up generating equivalent results.

Splitting the macros into 32/64 bit versions and avoiding a ?: with (uint32_t) on one leg and (uint64_t) on the other avoids making the overall result be uint64_t. I do have a replacement for the original `z_tmcvt` inline  which returned uint64_t, but that's only present to make a couple of tests work without changes. Maybe this should be removed though? It could easily be mis-used and cause 64-bit values to propagate beyond the time conversion operation.

As to generalizing these macros for other uses, that's certainly possible. We should find another use before attempting to do that as the results will end up a lot cleaner.

File: include/zephyr/sys/time_units.h [hunk 4]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
+	((__to_hz) == (__from_hz))
+
+/* true if the conversion requires a simple integer multiply */
+#define z_tmcvt_mul_ratio(__from_hz, __to_hz) \
+	((__to_hz) > (__from_hz) && (__to_hz) % (__from_hz) == 0U)
+
+/* true if the conversion requires a simple integer division */
+#define z_tmcvt_div_ratio(__from_hz, __to_hz) \
+	((__from_hz) > (__to_hz) && (uint64_t) (__from_hz) % (__to_hz) == 0U)
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:129]: > Out of interest: Why is the (uint64_t) cast required? AFAICS no risk of overflowing here.

Good catch. I've removed that.

> I think here and in other instances below some casts do just the same as automatic C type casting. This limits the general usability of the macros and maybe the compiler's leeway for optimization in case other input types will be used?

Agreed. I've removed some additional casts related to the from_hz and to_hz values but left all of the casts for the input value to uint64_t -- that ensures that computations will be wide enough to avoid overflows while still allowing the compiler to do whatever optimizations it can manage with the input and output frequency values.

> Actually the mul/div macros do the same and could be one (switching inputs). The naming of the macro could be improved then - some infix like "is_integer_divisible" or "is_multiple_of" would better capture the actual intention IMO.

Yup, these names came from the original source code, but as they had `bool` type, it was easier to see what they did. I've changed those to `z_tmcvt_is_identity`, `z_tmcvt_is_int_mul` and `z_tmcvt_is_int_div` to make it clear that they are predicates.

> Can be done in a separate commit such that verifying the correctness of the steps remains easy.

The existing tests have been very good at catching implementation errors; I'm going to squash these updates into the original patch as reading the differences is actually harder than just looking at the new code.

File: include/zephyr/sys/time_units.h [hunk 5]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
+
+/*
+ * Compute the offset needed to round the result correctly when
+ * the conversion requires a simple integer division
+ */
+#define z_tmcvt_off_div(__from_hz, __to_hz, __round_up, __round_off)	\
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:135]: Hrm. The `z_tmcvt_off_div` macro does help make sure the optimized 32-bit division case generates correct results as we need to range check the input time value against (1<<32) - offset. We could make that check less precise by simply checking for `(uint64_t)(__t) <= 0xffffffffU` and force some possibly optimizable instances through the slower path. That would certainly make the code a lot simpler.

In any case, reviewing this code did encourage me to switch from < BIT64(32) to <= 0xffffffffU -- that should keep computations in 32 bits more often.

File: include/zephyr/sys/time_units.h [hunk 6]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
+
+/* Slow 64-bit conversion for constant frequencies. This avoids overflowing the multiply */
+#define z_tmcvt_const_64_slow(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(((uint64_t) (__t) / (__from_hz))*(__to_hz) +			\
+	 (((uint64_t) (__t) % (__from_hz))*(__to_hz) +		\
+	  z_tmcvt_off_gen(__from_hz, __to_hz, __round_up, __round_off)) / (__from_hz))
+
+/* General 64-bit conversion for constant frequencies. Uses one of the two above macros */
+#define z_tmcvt_const_64(__t, __from_hz, __to_hz, __round_up, __round_off) \
+	(Z_TMCVT_USE_FAST_ALGO(__from_hz, __to_hz) ?			\
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:167]: Yup. That was from the original code, but switching case makes good sense.

File: include/zephyr/sys/time_units.h [hunk 7]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
+	 z_tmcvt_const_64_fast(__t, __from_hz, __to_hz, __round_up, __round_off) : \
+	 z_tmcvt_const_64_slow(__t, __from_hz, __to_hz, __round_up, __round_off))
+
+/* Functions for cases with non-constant frequencies */
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:171]: Yeah, I've actually gone ahead and renamed the macros to remove 'const' -- these macros don't actually depend upon having constant parameters. With that, I've removed the non-macro functions from timeutil.c and simply use the general macros for the non-constant case. Thanks for getting me to look at this more closely.

File: include/zephyr/sys/time_units.h [hunk 8]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
+uint32_t _z_tmcvt_32(uint64_t t, uint32_t from_hz,
+		     uint32_t to_hz, bool round_up,
+		     bool round_off);
+
+uint64_t _z_tmcvt_64(uint64_t t, uint32_t from_hz,
+		     uint32_t to_hz, bool round_up,
+		     bool round_off);
+
+/* Convert, generating a 32-bit result */
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:180]: Agreed. *and* using `uint64_t` for `__t` instead of having the _32 macros take in 32-bit values which might lead to truncation. I'm thinking particularly of the cycle to ms conversion where cycle counts are likely to be large while ms might well be within a 32-bit range. __t is *always* used as a 64-bit value now.

File: include/zephyr/sys/time_units.h [hunk 9]
```diff
@@ -111,70 +111,128 @@ static TIME_CONSTEXPR inline int sys_clock_hw_cycles_per_sec(void)
...
+		 (uint32_t) (__t)*((__to_hz) / (__from_hz))		\
+		 :							\
+		 z_tmcvt_const_32(__t, __from_hz, __to_hz, __round_up, __round_off) \
+		 )							\
+	 :								\
+	 _z_tmcvt_32(__t, __from_hz, __to_hz, __round_up, __round_off)	\
+		)
+
+/* Convert, generating a 64-bit result */
+#define z_tmcvt_64(__t, __from_hz, __to_hz, __const_hz, __round_up, __round_off) \
```
[the remark this answers is not in the document]
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:212]: Yeah, that's certainly possible -- a single macro with a parametric type. That's how the original code worked. I think with your idea of extracting the conversion computations to separate macros and using simple if_is_a ? do_a structure, the two final macros are pretty readable. I hate to introduce yet another layer of macro indirection into this process...

File: include/zephyr/sys/time_units.h (hunk 1 above)
Comment (keith-packard) [reply include/zephyr/sys/time_units.h:189]: Great idea. I've implemented this and the resulting `z_tmcvt_32` and `z_tmcvt_64` macros are pretty simple looking now.

File: include/zephyr/sys/time_units.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (fgrandel) [reply include/zephyr/sys/time_units.h:334]: I personally like it, especially as it is generated documentation which doesn't have to be maintained manually. Let's see what the others think. :-)

File: include/zephyr/sys/time_units.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (fgrandel) [reply include/zephyr/sys/time_units.h:220]: Agreed - with this decision and the others you took. The fact that you converted this from a function to a function generator will be immediately useful within #60400 - my own approach was much uglier. Looking forward to rebasing my stuff on top of it.

**Review by ghost — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Comment by keith-packard** [comment]
Anyone have an idea about how to debug the BabbleSim failures?

**Comment by fgrandel** [comment]
> Anyone have an idea about how to debug the BabbleSim failures?

Actually that BabbleSim errors are not so difficult to reproduce locally:

```console
$> west config manifest.group-filter -- +babblesim
$> cd tools/bsim
$> make everything -j 8
$> tests/bsim/bluetooth/host/att/eatt_notif/test_scripts/eatt_notif.sh # or whatever fails
```
(copied from my shell history - so might not be 100% correct/complete)

Then you should get nice logs that point you to the exact instruction that fails.

See https://babblesim.github.io/fetching.html, https://babblesim.github.io/building.html shamelessly copied from https://docs.zephyrproject.org/latest/develop/test/bsim.html.

**Comment by sylvioalves** [comment]
> oh, and there's one really ugly kludge left -- i had to leave the size of the init_delay variable set to 4 bytes or none of the tests for the m5stack_core2 would link. Something something overlapping linker segments. Anyone understand that platform enough to help out?

Having ELF check in `esptool` is welcome. However, should Zephyr has it proper aligned in Linker level as well? Or it is already?

**Comment by marekmatej** [comment]
> Turns out 'esptool' was ignoring alignment when merging ELF sections -- the 64-bit timer value caused the static thread init sections to require 8-byte instead of 4-byte alignment. PR pending at [zephyrproject-rtos/hal_espressif#243](https://github.com/zephyrproject-rtos/hal_espressif/pull/243)

@keith-packard  I remember having alignment problems with Zephyr LD scripts before in `include/zephyr/linker/device-deps.ld`, where alig(4) before the `__device_deps_end` helped to solve the issue. However, in the end, it was fixed otherwise. Maybe you can check other scripts for missing alignment too.

**Comment by fgrandel** [comment]
> It seems that more adventures await me...

Feel free to tag me if you need any assistance in the IEEE 802.15.4 area. Looks, though, that the failures are more OT related. Maybe @rlubos can give a hint?

**Comment by rlubos** [comment]
As for the BabbleSim failure (or the `echo_server` sample operation in general) this seems to be the culprit:
https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/sockets/echo_server/src/udp.c#L29

The sample specifies infinite timeout for the thread init delay, which no longer seems to work with static thread creation. So IPv4 thread starts regardless, causing the sample to fail (as with OT/15.4 L2 IPv4 is disabled).

Now the question is - shall infinite timeouts be supported with static thread creation? It used to work, but I actually can't find this behavior documented anywhere. If not, some samples/tests may need an update.

**Comment by keith-packard** [comment]
> As for the BabbleSim failure (or the `echo_server` sample operation in general) this seems to be the culprit: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/sockets/echo_server/src/udp.c#L29
> 
> The sample specifies infinite timeout for the thread init delay, which no longer seems to work with static thread creation. So IPv4 thread starts regardless, causing the sample to fail (as with OT/15.4 L2 IPv4 is disabled).
> 
> Now the question is - shall infinite timeouts be supported with static thread creation? It used to work, but I actually can't find this behavior documented anywhere. If not, some samples/tests may need an update.

Thanks 1<<20 for debugging this. I've switched the code from using `K_MSEC `to `SYS_TIMEOUT_MS` which checks for `SYS_FOREVER_MS` (which is -1).

**Comment by keith-packard** [comment]
> Having sections alignment check in `esptool` is welcome. However, should Zephyr has it proper alignment in Linker level as well? Or it is already?

The problem wasn't that the output wasn't aligned correctly (it was). The problem was that when esptool attempted to merge adjacent sections, it was surprised by the gaps between sections caused by alignment. The esptool patch checks for these gaps and pads the previous section when they occur so that the section merging code is happy.

**Comment by cfriedt** [comment]
@keith-packard - is this PR blocked? Looks like you can resolve conflicts, and if it's ok, we can remove the DNM?

**Comment by keith-packard** [comment]
> @keith-packard - is this PR blocked? Looks like you can resolve conflicts, and if it's ok, we can remove the DNM?

Yes, it's blocked waiting for the espressif module to pull in a fix for esptool that makes it deal with padding caused by alignment changes in this series. https://github.com/zephyrproject-rtos/hal_espressif/pull/243

**Comment by keith-packard** [comment]
> > @keith-packard - is this PR blocked? Looks like you can resolve conflicts, and if it's ok, we can remove the DNM?
> 
> Yes, it's blocked waiting for the espressif module to pull in a fix for esptool that makes it deal with padding caused by alignment changes in this series. [zephyrproject-rtos/hal_espressif#243](https://github.com/zephyrproject-rtos/hal_espressif/pull/243)

All fixed; west.yml now points at the updated espressif hal module.

**Review by cfriedt — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Review by andyross — APPROVED** [review]
Sure.   But note that the macro API still takes ms, so this isn't really doing anything in practice beyond removing the instructions to do the conversion at runtime (which obviously does have value).  IIRC this got left alone when the timeout rework happened because I had a vain hope of removing the idea of static thread timeouts in favor of something more orthogonal.  But obviously that never happened.

**Comment by evgeniy-paltsev** [comment]
@keith-packard - just FYI - this causing lot's of test failures on HW ARC platforms (its under debug RN). So this still have some flaws.

**Comment by keith-packard** [comment]
> @keith-packard - just FYI - this causing lot's of test failures on HW ARC platforms (its under debug RN). So this still have some flaws.

Let me know if I can help in some way. Not that I have any ARC hardware, but ...

**Comment by kokas-a** [comment]
Hi @keith-packard!
As @evgeniy-paltsev has mentioned, we have a lot of faults in our tests after applaying https://github.com/zephyrproject-rtos/zephyr/commit/8f5d0791bfc5c1a8be8d98e4b838ccd634fdc6db commit.
In our test configuration we use ARC-based hardware and LLVM-based ARCMWDT compiler.
I've tryed to debug this faults, but I still unable to findout the rootcase of the problem. I made such observation:
If I make call macroses with help of inline functions (https://github.com/zephyrproject-rtos/zephyr/commit/6ca2324cb94ada0b40865a02f4132e860aedf168) all tests become passes.
In simplest tests assembly listings of both cases (macroses only and macroses wrapped with inline functions) are the same. 
But in common case something went wrong. As for me it looks like overoptimisation on compiler side and such behavior happends in some cases when ```volatile``` modifier is ommited for some variable. 
Based on such description could you make a suggestion what is wrong?

**Comment by keith-packard** [comment]
> Hi @keith-packard! As @evgeniy-paltsev has mentioned, we have a lot of faults in our tests after applaying [8f5d079]

Where are you seeing faults? Have you tried with clang fix applied (which v3.5.0-rc2 contains)? Have you tried with 
41e8b44619c7dd60ccf91ac642c49e68d14f7ede reverted?

>But in common case something went wrong. 

Can you identify a specific case which gets mis-compiled and compare before/after compiler outputs?

> As for me it looks like overoptimisation on compiler side and such behavior happends in some cases when `volatile` modifier is ommited for some variable. Based on such description could you make a suggestion what is wrong?

Volatile isn't relevant here -- there are no variables involved as the expressions are all constant and should be evaluated at compile time.
